lxd/storage/drivers/drivers_lvm.go

```diff
@@ -0,0 +1,532 @@
+		}
+		defer loopFile.Close()
+	}
+
+	vgExists, vgTags, err := d.volumeGroupExists(d.config["lvm.vg_name"])
+	if err != nil {
+		return err
+	}
+
+	removeVg := false
+	if vgExists {
+		// Count normal and thin volumes.
+		lvCount, err := d.countLogicalVolumes(d.config["lvm.vg_name"])
+		if err != nil && err != errLVMNotFound {
+			return err
+		}
+
+		// Check that volume group is not in use. If it is we need to assume that other users are using
+		// the volume group, so don't remove it. This actually goes against policy since we explicitly
+		// state: our pool, and nothing but our pool, but still, let's not hurt users.
+		if err == nil {
+			if lvCount == 0 {
+				removeVg = true // Volume group is totally empty, safe to remove.
+			} else if d.usesThinpool() && lvCount > 0 {
+				// Lets see if the lv count is just our thin pool, or whether we can only remove
+				// the thin pool itself and not the volume group.
+				thinVolCount, err := d.countThinVolumes(d.config["lvm.vg_name"], d.thinpoolName())
+				if err != nil && err != errLVMNotFound {
+					return err
+				}
+
+				// Thin pool exists.
+				if err == nil {
+					// If thin pool is empty and the total VG volume count is 1 (our thin pool
+					// volume) then just remote the entire volume group.
+					if thinVolCount == 0 && lvCount == 1 {
+						removeVg = true
+					} else if thinVolCount == 0 && lvCount > 1 {
+						// Otherwise, if the thin pool is empty but the volume group has
+						// other volumes, then just remove the thin pool volume.
+						err = d.removeLogicalVolume(d.lvmDevPath(d.config["lvm.vg_name"], "", "", d.thinpoolName()))
+						if err != nil {
+							return errors.Wrapf(err, "Failed to delete thin pool %q from volume group %q", d.thinpoolName(), d.config["lvm.vg_name"])
+						}
+						d.logger.Debug("Thin pool removed", log.Ctx{"vg_name": d.config["lvm.vg_name"], "thinpool_name": d.thinpoolName()})
+					}
+				}
+			}
+		}
+
+		// Remove volume group if needed.
+		if removeVg {
+			_, err := shared.TryRunCommand("vgremove", "-f", d.config["lvm.vg_name"])
+			if err != nil {
+				return errors.Wrapf(err, "Failed to delete the volume group for the lvm storage pool")
+			}
+			d.logger.Debug("Volume group removed", log.Ctx{"vg_name": d.config["lvm.vg_name"]})
+		} else {
+			// Otherwise just remove the lvmVgPoolMarker tag to indicate LXD no longer uses this VG.
+			if shared.StringInSlice(lvmVgPoolMarker, vgTags) {
+				_, err = shared.TryRunCommand("vgchange", "--deltag", lvmVgPoolMarker, d.config["lvm.vg_name"])
+				if err != nil {
+					return errors.Wrapf(err, "Failed to remove marker tag on volume group for the lvm storage pool")
+				}
+				d.logger.Debug("LXD marker tag removed from volume group", log.Ctx{"vg_name": d.config["lvm.vg_name"]})
+			}
+		}
+	}
+
+	// If we have removed the volume group and this is a loop file, lets clean up the physical volume too.
+	if removeVg && loopFile != nil {
+		err = SetAutoclearOnLoopDev(int(loopFile.Fd()))
```

Comment:
@stgraber I've reinstated this part after I removed it this morning after your suggestion to call `d.Unmount()` and add a call to `releaseLoopDev` instead.

This is because I need to use `openLoopFile` to get the loop files's device address in order to run `pvremove` against it, however also the `releaseLoopDev` in `Unmount` didn't work reliably (perhaps there is a race in pvremove that meant that it still had a file handle open to the device preventing it from working).

By setting the auto clear flag on the device, once everything has close the device it is auto detached.

Reply:
There's no need to call `Unmount` here because the vg will already have been removed.